pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +234,239 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+
+		// Add existing ciImports to WireGuard Peers
+		return c.updateWireGuardPeers("")
+	}
+
+	return nil
+}
+
+func (c *MCDefaultRouteController) deleteWireGuardInterfaceAndRule() error {
+	if err := c.routeClient.DeleteIPRule(multiclusterWireGuardFwmark, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	return c.wireGuardClient.Cleanup()
+}
+
+func (c *MCDefaultRouteController) initWireGuardRoute() error {
+	if err := c.routeClient.AddIPRule(multiclusterWireGuardFwmark, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	if err := c.routeClient.AddRouteToTable(multiclusterWireGuardInterface, net.IPv4zero, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MCDefaultRouteController) updateWireGuardPeers(ciImp string) error {
+	c.ciImportMutex.Lock()
+	defer c.ciImportMutex.Unlock()
+	if !c.wireGuardInitialized {
+		return fmt.Errorf("failed to update WireGuard peers, the WireGuard has not been initialized")
+	}
+
+	if ciImp != "" {
+		ciImport, err := c.ciImportLister.ClusterInfoImports(c.namespace).Get(ciImp)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			klog.InfoS("The ClusterInfoImport has been deleted, removing WireGuard peer", "ClusterInfoImport", ciImp)
+			return c.wireGuardClient.DeletePeer(ciImp)
+		}
+		return c.updateSinglePeer(ciImport)
+	}
+
+	// The ClusterInfoImport's name is empty. It means that a new Gateway is created, controller should add all existing ClusterInfoImport
+	// to WireGuard peers.
+	ciImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, ciImport := range ciImports {
+		if err := c.updateSinglePeer(ciImport); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MCDefaultRouteController) updateSinglePeer(ciImport *mcv1alpha1.ClusterInfoImport) error {
+	gatewayIP := net.ParseIP(ciImport.Spec.GatewayInfos[0].GatewayIP)
+	_, gatewayIPCIDR, err := net.ParseCIDR(ciImport.Spec.GatewayInfos[0].GatewayIP + "/32")
+	if err != nil {
+		return err
+	}
+	_, serviceIPCIDR, err := net.ParseCIDR(ciImport.Spec.ServiceCIDR)
+	if err != nil {
+		return err
+	}
+	podCIDRs := []*net.IPNet{}
+	for _, cidr := range ciImport.Spec.PodCIDRs {
+		_, podCIDR, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return err
+		}
+		podCIDRs = append(podCIDRs, podCIDR)
+	}
+
+	allowedIPs := []*net.IPNet{serviceIPCIDR, gatewayIPCIDR}
+	allowedIPs = append(allowedIPs, podCIDRs...)
+	if ciImport.Spec.WireGuard != nil && ciImport.Spec.WireGuard.PublicKey != "" {
+		if err := c.wireGuardClient.UpdatePeer(ciImport.Name, ciImport.Spec.WireGuard.PublicKey, gatewayIP, 0, allowedIPs); err != nil {
+			return err
+		}
+		klog.InfoS("Updated WireGuard peer with ClusterInfoImport", "ClusterInfoImport", ciImport)
+	}
+	return nil
+}
+
+func (c *MCDefaultRouteController) initializeWireGuard() error {
+	c.networkConfig.CalculateMTUDeduction(false)
```

Comment:
Do not call `CalculateMTUDeduction()` again, which will change `networkConfig.MTUDeduction`. Could not we just use `networkConfig.MTUDeduction` which should have been set?